From 8295ec8473a979e69943ef4fccf197c16dc710ef Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Fri, 12 Sep 2014 11:48:40 -0400
Subject: [ethernet] i40e: Allow user to change link settings if link is down

Message-id: <1410522528-19793-3-git-send-email-sassmann@redhat.com>
Patchwork-id: 93935
O-Subject: [RHEL7.1 PATCH 02/10] i40e: Allow user to change link settings if link is down
Bugzilla: 1140994
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

From: Catherine Sullivan <catherine.sullivan@intel.com>

Allow the user to change auto-negotiation and speed settings if
link is down.

Change-ID: I372967c627682b5e1835f623a7cbf41b21b51043
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Jim Young <jamesx.m.young@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit c57e9f179b5cd2f4fbdfc973e9174094b06ca43b)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index f3a0117..f0ab3e7 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -466,7 +466,8 @@ static int i40e_set_settings(struct net_device *netdev,
 
  if (hw->phy.media_type != I40E_MEDIA_TYPE_BASET &&
      hw->phy.media_type != I40E_MEDIA_TYPE_FIBER &&
-     hw->phy.media_type != I40E_MEDIA_TYPE_BACKPLANE)
+     hw->phy.media_type != I40E_MEDIA_TYPE_BACKPLANE &&
+     hw->phy.link_info.link_info & I40E_AQ_LINK_UP)
   return -EOPNOTSUPP;
 
  /* get our own copy of the bits to check against */
-- 
1.7.1